src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java

```diff
@@ -248,37 +258,95 @@ public void onResume() {
+          PauseAction.endCurrentPause(getContextClass(FlowNode.class));
           SonarQubeWebHook.get().removeListener(this);
-          log("SonarQube task '%s' status is '%s'", step.taskId, taskStatus);
-          switch (taskStatus) {
-            case WsClient.CETask.STATUS_SUCCESS:
-              log("SonarQube task '%s' completed. Quality gate is '%s'", step.taskId, qgStatus);
-              handleQGStatus(qgStatus);
-              break;
-            case WsClient.CETask.STATUS_FAILURE:
-            case WsClient.CETask.STATUS_CANCELED:
-              getContext().onFailure(new IllegalStateException("SonarQube analysis '" + step.getTaskId() + "' failed: " + taskStatus));
-              break;
-            default:
-              throw new IllegalStateException("Unexpected task status: " + taskStatus);
+          if (step.webhookSecretId != null) {
+            if (validateWebhook(receivedSignature, payload.getPayloadAsString())) {
+              // only execute the checkQualityGate if the webhook is found to be valid (getContext().onFailure() does not interrupt execution)
+              checkQualityGate(payload.getTaskStatus(), payload.getQGStatus());
+            }
+          } else {
+            checkQualityGate(payload.getTaskStatus(), payload.getQGStatus());
           }
-        } catch (IOException | InterruptedException e) {
-          LOGGER.log(Level.WARNING, "Error during WaitForQualityGateStep", e);
+        } catch (IOException e) {
+          getContext().onFailure(e);
+          throw new IllegalStateException(e);
         }
       }
     }
+
+    private void checkQualityGate(String taskStatus, @Nullable String qgStatus) {
+      log("SonarQube task '%s' status is '%s'", step.taskId, taskStatus);
+      switch (taskStatus) {
+        case WsClient.CETask.STATUS_SUCCESS:
+          log("SonarQube task '%s' completed. Quality gate is '%s'", step.taskId, qgStatus);
+          handleQGStatus(qgStatus);
+          break;
+        case WsClient.CETask.STATUS_FAILURE:
+        case WsClient.CETask.STATUS_CANCELED:
+          getContext().onFailure(new IllegalStateException("SonarQube analysis '" + step.getTaskId() + "' failed: " + taskStatus));
+          break;
+        default:
+          throw new IllegalStateException("Unexpected task status: " + taskStatus);
+      }
+    }
+
+    private boolean validateWebhook(String receivedSignature, String payload) {
+      StringCredentials webhookSecret = CredentialsProvider.findCredentialById(step.webhookSecretId, StringCredentials.class, getContextClass(Run.class));
+      CredentialsProvider.track(getContextClass(Run.class), webhookSecret);
+      if (webhookSecret != null) {
+        boolean isValidPayload = isValidSignature(receivedSignature, payload, webhookSecret.getSecret().getPlainText());
+        if (!isValidPayload) {
+          log("The incoming webhook was verified using the configured webhook secret and was found to be invalid");
```

Comment:
I find this phrasing a bit long. How about:

```suggestion
          log("The incoming webhook does not match the configured secret");
```

Reply:
That does sound better